### PR TITLE
Better formatting for MSSQL when using National Characters

### PR DESCRIPTION
**Pull request by mtxr**
formatting queries with `N'value here'`.

### CODE REVIEW

**Comment by nene** [comment]
So, this handles the `N'string'`, though when you format a query like:

```
SELECT N'some string';
```

You'd expect a formatting like:

```
SELECT
  N'some string';
```

But instead the formatter now produces:

```
SELECT N'some string';
```

Also, when given a query like:

```
SELECT N, M FROM MY_TABLE;
```

You'd expect a result like:

```
SELECT
  N,
  M
FROM
  MY_TABLE;
```

But instead we get:

```
SELECT N,
  M
FROM
  MY_TABLE;
```

I think it would be better to handle this special string syntax inside Tokenizer, where all other strings are detected. Also, all other string syntaxes are configurable, so should be this. So we could configure strings like:

```
stringTypes: ["''", "N''"]
```

File: test/StandardSqlFormatterTest.js [hunk 1]
```diff
@@ -202,6 +202,67 @@ describe("StandardSqlFormatter", function() {
...
 
+    it("formats simple SELECT", function() {
+        const result = sqlFormatter.format("SELECT 'value'", );
+        expect(result).toBe(
+            "SELECT\n" +
+            "  'value'\n"
+        );
+    });
+
+    it("formats simple SELECT", function() {
```
Comment (ukupat) [inline test/StandardSqlFormatterTest.js]: It has a duplicate test description. Could we be more specific about why is this test different from the above one?

File: src/core/Tokenizer.js [hunk 2]
```diff
@@ -54,10 +54,12 @@ export default class Tokenizer {
     // 2. square bracket quoted string (SQL Server) using ]] to escape
     // 3. double quoted string using "" or \" to escape
     // 4. single quoted string using '' or \' to escape
+    // 5. national character single quoted string using N'' or \' to escape
```
Comment (ukupat) [inline src/core/Tokenizer.js]: Great comment! `constructor` `cfg.stringTypes` comment needs also some updating.

File: src/core/Formatter.js [hunk 3]
```diff
@@ -165,7 +165,7 @@ export default class Formatter {
     }
 
     formatWithSpaceAfter(token, query) {
-        return _.trimEnd(query) + token.value + " ";
+        return this.formatWithoutSpaces(token, _.trimEnd(query)) + " ";
```
Comment (ukupat) [inline src/core/Formatter.js]: Why did we change this? If we want to use `formatWithoutSpaces` to reduce duplication then we could use it like this: 

```
return this. formatWithoutSpaces(token, query) + " ";
```

 But I think it was easier to understand `formatWithSpaceAfter` logic when there was a bit of duplication in the code.

**Comment by ukupat** [comment]
@mtxr – waiting for your changes.

**Comment by mtxr** [comment]
Sorry. I missed this notification, I'll do this by the end of the day

File: test/StandardSqlFormatterTest.js (hunk 1 above)
Comment (mtxr) [inline test/StandardSqlFormatterTest.js]: Fixed on mtxr/sql-formatter@d6a294dc8b1c5e2a3c1b2e5f93488b265e9de78e

File: src/core/Tokenizer.js (hunk 2 above)
Comment (mtxr) [inline src/core/Tokenizer.js]: Thanks!

File: src/core/Formatter.js (hunk 3 above)
Comment (mtxr) [inline src/core/Formatter.js]: Fixed on mtxr/sql-formatter@d6a294dc8b1c5e2a3c1b2e5f93488b265e9de78e

File: test/behavesLikeSqlFormatter.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ export default function behavesLikeSqlFormatter(language) {
     it("uses given indent config for indention", function() {
         const result = sqlFormatter.format(
             "SELECT count(*),Column1 FROM Table1;",
-            {language, indent: "    "}
+            { language, indent: "    " }
```
Comment (ukupat) [inline test/behavesLikeSqlFormatter.js]: This change is not needed.

File: test/StandardSqlFormatterTest.js [hunk 5]
```diff
@@ -202,6 +202,66 @@ describe("StandardSqlFormatter", function() {
         );
     });
 
+    it("formats simple SELECT", function() {
```
Comment (ukupat) [inline test/StandardSqlFormatterTest.js]: This case is already covered by other tests, therefore we can remove it.

File: test/StandardSqlFormatterTest.js [hunk 6]
```diff
@@ -202,6 +202,66 @@ describe("StandardSqlFormatter", function() {
...
+        const result = sqlFormatter.format(
+            "SELECT N'value';"
+        );
+        expect(result).toBe(
+            "SELECT\n" +
+            "  N'value';\n"
+        );
+    });
+
+    it("formats SELECT with complex WHERE", function() {
```
Comment (ukupat) [inline test/StandardSqlFormatterTest.js]: This test file is getting quite huge. We could save some lines when merge this test and below one:

```
it("formats SELECT with complex WHERE", function() {
    const result = sqlFormatter.format(
        "SELECT * FROM foo WHERE Column1 = N'testing' AND Column2 = 'testing'" +
        "AND ( (Column3 = Column4 OR Column5 >= NOW()) );"
    );
    expect(result).toBe(
        "SELECT\n" +
        "  *\n" +
        "FROM\n" +
        "  foo\n" +
        "WHERE\n" +
        "  Column1 = N'testing'\n" +
        "  AND (\n" +
        "  Column2 = 'testing'\n" +
        "    (\n" +
        "      Column3 = Column4\n" +
        "      OR Column5 >= NOW()\n" +
        "    )\n" +
        "  );\n"
    );
});
```

File: src/core/Tokenizer.js (hunk 2 above)
Comment (ukupat) [inline src/core/Tokenizer.js]: I think you forgot to update `Tokenizer` `constructor` `cfg.stringTypes` comment.

Comment (mtxr) [inline src/core/Tokenizer.js]: `5. national character quoted string using N'' or N\' to escape`

Is it ok?

Comment (ukupat) [inline src/core/Tokenizer.js]: That is cool, but line 10 needs update as well :)

**Comment by ukupat** [comment]
@mtxr – here you go https://github.com/zeroturnaround/sql-formatter/releases/tag/v1.2.2

**Pull request merged**